### PR TITLE
Refactor pravega writer and readers with the abstract Writers and readers

### PR DESCRIPTION
**Pull request by kmgowda**
The new abstract classes WriteWorker and ReadWorkers are created.
These classes does the performance bench-marking whereas the
pravega writer and pravega readers classes does the data writes and reads.

### CODE REVIEW

File: src/main/java/com/emc/pravega/perf/PravegaReaderWorker.java [hunk 1]
```diff
@@ -30,60 +28,44 @@
...
-                if (ret != null) {
-                    stats.recordTime(null, Instant.ofEpochMilli(Long.parseLong(ret.split(",")[0])), ret.length());
-                }
-            } while ((Duration.between(StartTime, Instant.now()).getSeconds() < secondsToRun) &&
-                    (eventCount.incrementAndGet() < totalEvents));
-        } finally {
-            reader.close();
+            return reader.readNextEvent(timeout).getEvent();
+        } catch(ReinitializationRequiredException e){
+            throw new Exception("ReinitializationRequiredException exception while reading data ");
```
Comment (tkaitchuck) [inline src/main/java/com/emc/pravega/perf/PravegaReaderWorker.java]: This doesn't seem useful.

File: src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java [hunk 2]
```diff
@@ -34,30 +34,33 @@
...
+                messageSize, start, stats, streamName, totalEvents, factory);
 
         this.transactionsPerCommit = transactionsPerCommit;
         eventCount = new AtomicInteger(0);
         transaction = new AtomicReference<Transaction<String>>(producer.beginTxn());
     }
 
     @Override
-    public CompletableFuture writeData(String key, String data) throws TxnFailedException, IllegalStateException {
+    public CompletableFuture writeData(String key, String data) throws IllegalStateException {
```
Comment (tkaitchuck) [inline src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java]: IllegalStateException is unchecked.

File: src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java [hunk 3]
```diff
@@ -34,30 +34,33 @@
...
-            eventCount.set(0);
-            curTrans.commit();
-            if (!transaction.compareAndSet(curTrans, producer.beginTxn())) {
-                throw new IllegalStateException("WriteData called on the same PravegaTransactionWriterWorker from two threads in parallel.");
+        try {
+            curTrans.writeEvent(key, data);
+            if (eventCount.incrementAndGet() >= transactionsPerCommit) {
+                eventCount.set(0);
+                curTrans.commit();
+                if (!transaction.compareAndSet(curTrans, producer.beginTxn())) {
```
Comment (tkaitchuck) [inline src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java]: This logic is not thread safe.

File: src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java [hunk 4]
```diff
@@ -34,30 +34,33 @@
...
+                    throw new IllegalStateException("WriteData called on the same PravegaTransactionWriterWorker from two threads in parallel.");
+                }
             }
+        } catch (TxnFailedException e) {
+            throw new IllegalStateException("Write data failed with transaction failure.");
```
Comment (tkaitchuck) [inline src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java]: I'm not sure this qualfies as an illegal state.

File: src/main/java/com/emc/pravega/perf/PravegaWriterWorker.java [hunk 5]
```diff
@@ -32,82 +29,41 @@
...
-
+    public void flush() {
         producer.flush();
-        // producer.close();
+    }
 
-        //Wait for the last packet to get acked
-        retFuture.get();
+    @Override
+    public long eventCountIncrementAndGet() {
```
Comment (tkaitchuck) [inline src/main/java/com/emc/pravega/perf/PravegaWriterWorker.java]: Why is this a public method?

File: src/main/java/com/emc/pravega/perf/ReaderWorker.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+    public abstract String readData() throws Exception;
+
+    public abstract void close();
+
+    public abstract long eventCountIncrementAndGet();
+
+    public abstract long eventCountGet();
+
+    public void cleanupEvents(PerfStats drainStats) throws Exception {
```
Comment (tkaitchuck) [inline src/main/java/com/emc/pravega/perf/ReaderWorker.java]: All of these public methods and those elsewhere need to have Javadocs. 
For example I have no idea why "cleanupEvents" would involve invoking "readData".

File: src/main/java/com/emc/pravega/perf/WriterWorker.java [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    public abstract CompletableFuture writeData(String key, String data);
+
+    public abstract void flush();
+
+    public abstract long eventCountIncrementAndGet();
+
+    public abstract long eventCountGet();
+
+    @Override
+    public Void call() throws InterruptedException, ExecutionException, IllegalStateException {
```
Comment (tkaitchuck) [inline src/main/java/com/emc/pravega/perf/WriterWorker.java]: Here and elsewhere: You don't need to declare runtime exception in the signature.

File: src/main/java/com/emc/pravega/perf/WriterWorker.java [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+            final Instant loopStartTime = Instant.now();
+            for (int i = 0; (i < eventsPerSec) && (eventCountIncrementAndGet() <= totalEvents) &&
+                    (Duration.between(loopStartTime, Instant.now()).getSeconds() < 1); i++) {
+
+                // Construct event payload
+                String val = System.currentTimeMillis() + ", " + workerID + ", " + (int) (Math.random() * 200);
+                String payload = String.format("%-" + messageSize + "s", val);
+                String key;
+                if (isRandomKey) {
+                    key = Integer.toString(workerID + new Random().nextInt());
```
Comment (tkaitchuck) [inline src/main/java/com/emc/pravega/perf/WriterWorker.java]: Don't use random as a one off like this. It is not random. You need to same it as a variable outside the loop so you can use it over and over. 
Also spotbugs should have caught this. Is it not running on this repo? If so we should enable it.

File: src/main/java/com/emc/pravega/perf/WriterWorker.java [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+
+                final Instant startTime = Instant.now();
+                retFuture = writeData(key, payload);
+                // event ingestion
+                retFuture = stats.recordTime(retFuture, startTime, payload.length());
+            }
+
+            long timeSpent = Duration.between(loopStartTime, Instant.now()).toMillis();
+            if (timeSpent < 1000) {
+                Thread.sleep(1000 - timeSpent);
```
Comment (tkaitchuck) [inline src/main/java/com/emc/pravega/perf/WriterWorker.java]: Please file an issue on dealing with over sleeps.

File: src/main/java/com/emc/pravega/perf/PravegaReaderWorker.java [hunk 10]
```diff
@@ -30,60 +28,44 @@
 import io.pravega.client.stream.EventRead;
 import io.pravega.client.stream.ReinitializationRequiredException;
 
-public class PravegaReaderWorker implements Callable<Void> {
+public class PravegaReaderWorker extends ReaderWorker  {
     public final static AtomicInteger eventCount = new AtomicInteger(0);
```
Comment (tkaitchuck) [inline src/main/java/com/emc/pravega/perf/PravegaReaderWorker.java]: I don't think this should be public or static.

File: src/main/java/com/emc/pravega/perf/PravegaReaderWorker.java [hunk 11]
```diff
@@ -30,60 +28,44 @@
...
-        return null;
     }
+
+    @Override
+    public void close(){
+        reader.close();
+    }
+
+    @Override
+    public long eventCountIncrementAndGet() {
```
Comment (tkaitchuck) [inline src/main/java/com/emc/pravega/perf/PravegaReaderWorker.java]: This should not be public

File: src/main/java/com/emc/pravega/perf/WriterWorker.java [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+        super(sensorId, eventsPerSec, secondsToRun,
+                isRandomKey, messageSize, start,
+                stats, streamName, totalEvents, 0);
+    }
+
+    public abstract CompletableFuture writeData(String key, String data);
+
+    public abstract void flush();
+
+    public abstract long eventCountIncrementAndGet();
```
Comment (tkaitchuck) [inline src/main/java/com/emc/pravega/perf/WriterWorker.java]: This should not be public

File: src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java [hunk 13]
```diff
@@ -34,30 +34,33 @@
...
+                                   PerfStats stats, String streamName,
+                                   long totalEvents, ClientFactory factory,
                                    int transactionsPerCommit) {
 
         super(sensorId, eventsPerSec, secondsToRun, isRandomKey,
-                messageSize, start, factory, stats, streamName, totalEvents);
+                messageSize, start, stats, streamName, totalEvents, factory);
 
         this.transactionsPerCommit = transactionsPerCommit;
         eventCount = new AtomicInteger(0);
```
Comment (tkaitchuck) [inline src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java]: This is masking eventCount for the super class

File: src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java [hunk 14]
```diff
@@ -34,30 +34,33 @@
     PravegaTransactionWriterWorker(int sensorId, int eventsPerSec,
```
Comment (tkaitchuck) [inline src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java]: I think this class could just extend WriterWorker directly.

File: src/main/java/com/emc/pravega/perf/PravegaReaderWorker.java (hunk 1 above)
Comment (kmgowda) [reply src/main/java/com/emc/pravega/perf/PravegaReaderWorker.java]: i need your help here Tom,   this what my design in brief.
WriterWorker  - An abstract class which implements call method the performance benchmarking.
                           This class will not know anything about pravega or any other streaming apis (like kafka)

PravegaWriterWorker  - Pravega specific writer class for writer benchmarking. This class extends  "WriterWorker  " .. 

so, here is the problem 

methods of PravegaWriterWorker class can throw pravega specific exceptions but  call method (for performance bench-marking) or any other abstract methods of the base abstract class WriterWorker  will not know about it.  So, here WriterWorker   should be generic.
hence, ReinitializationRequiredException  e is converted as generic exception.

please let me know if you want better of dealing the above scenario.

File: src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java (hunk 2 above)
Comment (kmgowda) [reply src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java]: same as above, since the writedata is abstract method of base class WriterWorker  , this class does not know TxnFailedException of pravega.

File: src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java (hunk 3 above)
Comment (kmgowda) [reply src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java]: eventCount is changed to transEventCount its Atomic interger and it increments rather than decrementing as per your last review comments. please let me know where exactly you are seeing an issue

File: src/main/java/com/emc/pravega/perf/PravegaWriterWorker.java (hunk 5 above)
Comment (kmgowda) [reply src/main/java/com/emc/pravega/perf/PravegaWriterWorker.java]: this is overriding the abstract method fromt the abstract class WriterWorker

File: src/main/java/com/emc/pravega/perf/PravegaReaderWorker.java (hunk 10 above)
Comment (kmgowda) [reply src/main/java/com/emc/pravega/perf/PravegaReaderWorker.java]: thanks.. I have made this one as private now

File: src/main/java/com/emc/pravega/perf/WriterWorker.java (hunk 12 above)
Comment (kmgowda) [reply src/main/java/com/emc/pravega/perf/WriterWorker.java]: no, this is an abstract methods must be overridden in derived class , should be public.

File: src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java (hunk 13 above)
Comment (kmgowda) [reply src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java]: thanks a lot Tom, I have changed this to transEventCount

File: src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java (hunk 14 above)
Comment (kmgowda) [reply src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java]: no, here the PravegaTransactionWriterWorker need the producer created by PravegaWriterWorker.

File: src/main/java/com/emc/pravega/perf/ReaderWorker.java (hunk 6 above)
Comment (kmgowda) [reply src/main/java/com/emc/pravega/perf/ReaderWorker.java]: yes, I have added Javdoc tags

File: src/main/java/com/emc/pravega/perf/WriterWorker.java (hunk 8 above)
Comment (kmgowda) [reply src/main/java/com/emc/pravega/perf/WriterWorker.java]: I will do this as separate PR please.. I need to lot of stuff during random key till then i am just keeping the existing code

File: src/main/java/com/emc/pravega/perf/WriterWorker.java (hunk 9 above)
Comment (kmgowda) [reply src/main/java/com/emc/pravega/perf/WriterWorker.java]: sure..  but what is the issue here? this seems ok for me.. if the events are pushed before time, then we can wait for next sec to start the writes right?

File: src/main/java/com/emc/pravega/perf/PravegaReaderWorker.java (hunk 11 above)
Comment (kmgowda) [reply src/main/java/com/emc/pravega/perf/PravegaReaderWorker.java]: no, this is an abstract methods must be overridden in derived class , need to be public.

File: src/main/java/com/emc/pravega/perf/PravegaReaderWorker.java (hunk 1 above)
Comment (tkaitchuck) [reply src/main/java/com/emc/pravega/perf/PravegaReaderWorker.java]: It's fine to convert one exception type to another specific type. But here it's just converting to "Exception" which does not serve any purpose because `ReinitializationRequiredException` already is an Exception. So this rethrowing could be deleted and it wouldn't affect the code in any way besides the string "ReinitializationRequiredException exception while reading data" being printed if the caller logs the exception. 
If you want to create some sort of WorkerFailureException then a conversion might make sense.

File: src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java (hunk 2 above)
Comment (tkaitchuck) [reply src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java]: yes, but IllegalState is unchecked. I does not need to be on this line. You can still throw it, there is just no need to type it here.

File: src/main/java/com/emc/pravega/perf/WriterWorker.java (hunk 8 above)
Comment (tkaitchuck) [reply src/main/java/com/emc/pravega/perf/WriterWorker.java]: Yes, add spotbugs as a separate PR. But please move the construction of random up to line 51.

File: src/main/java/com/emc/pravega/perf/WriterWorker.java (hunk 9 above)
Comment (tkaitchuck) [reply src/main/java/com/emc/pravega/perf/WriterWorker.java]: sleep is not guarenteed to wake up after the duration specified. It will wake up some time later. So what happens when you ask it to sleep for 300ms and it sleeps for 3s?

File: src/main/java/com/emc/pravega/perf/WriterWorker.java (hunk 12 above)
Comment (tkaitchuck) [reply src/main/java/com/emc/pravega/perf/WriterWorker.java]: abstract methods do not have to be public.

File: src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java (hunk 14 above)
Comment (tkaitchuck) [reply src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java]: Performing transaction on a Producer is a deprecated API. The new API works by creating a transactional producer from the client factory directly.

File: src/main/java/com/emc/pravega/perf/PravegaReaderWorker.java [hunk 15]
```diff
@@ -30,60 +28,48 @@
...
 
-public class PravegaReaderWorker implements Callable<Void> {
-    public final static AtomicInteger eventCount = new AtomicInteger(0);
-    public final long totalEvents;
+
+/**
+ *  class for Pravega reader/consumer.
+ */
+public class PravegaReaderWorker extends ReaderWorker  {
+    private final static AtomicInteger eventCount = new AtomicInteger(0);
```
Comment (tkaitchuck) [inline src/main/java/com/emc/pravega/perf/PravegaReaderWorker.java]: Making this static means that two instances of this test can never be run in the the same VM. That's a really odd restriction to impose, especially without documentation. Wouldn't a target rate an duration be sufficient to control when to exit?

File: src/main/java/com/emc/pravega/perf/ReaderWorker.java [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+    public Void call() throws ExecutionException, Exception {
+        String ret = null;
+        try {
+            do {
+                ret = readData();
+                if (ret != null) {
+                    stats.recordTime(null, Instant.ofEpochMilli(Long.parseLong(ret.split(",")[0])), ret.length());
+                }
+            } while ((Duration.between(StartTime, Instant.now()).getSeconds() < secondsToRun) &&
+                    (eventCountIncrementAndGet() < totalEvents));
```
Comment (tkaitchuck) [inline src/main/java/com/emc/pravega/perf/ReaderWorker.java]: eventCountIncrementAndGet() is an instance method that is defined above but it is being compared against a totalEvents count. 
If this is meant to be an instance variable, then it can simply be made into a local variable because it is not used in the child class.
If it is meant to be a global, then why is the child forced to implement it? there is only one possible implementation.

File: src/main/java/com/emc/pravega/perf/WriterWorker.java [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+     */
+    public abstract long eventCountGet();
+
+    @Override
+    public Void call() throws InterruptedException, ExecutionException, IllegalStateException {
+        CompletableFuture retFuture = null;
+
+        do {
+            final Instant loopStartTime = Instant.now();
+            for (int i = 0; (i < eventsPerSec) && (eventCountIncrementAndGet() <= totalEvents) &&
```
Comment (tkaitchuck) [inline src/main/java/com/emc/pravega/perf/WriterWorker.java]: Why do the children need to implement this?
eventCountIncrementAndGet() is an instance method that is defined above but it is being compared against a totalEvents count.
If this is meant to be an instance variable, then it can simply be made into a local variable because it is not used in the child class.
If it is meant to be a global, then why is the child forced to implement it? there is only one possible implementation.

File: src/main/java/com/emc/pravega/perf/PravegaReaderWorker.java (hunk 1 above)
Comment (kmgowda) [reply src/main/java/com/emc/pravega/perf/PravegaReaderWorker.java]: yes, then in this case ReinitializationRequiredException can be converted to illegalstateExcpetion

File: src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java (hunk 2 above)
Comment (kmgowda) [reply src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java]: i dint get this.. it throws IllegalStateException  correct?

File: src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java (hunk 3 above)
Comment (kmgowda) [reply src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java]: yes, I have moved  below two lines separate synchronized function: 
   
 private synchronized boolean transIncrementAndCompare() {
        if (transEventCount.incrementAndGet() >= transactionsPerCommit) {
            transEventCount.set(0);
            return true;
        }
        return false;
    }

File: src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java (hunk 4 above)
Comment (kmgowda) [reply src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java]: yes, I do feel so.. but can you suggest which generic exception i can use? shall i use RuntimeException?

File: src/main/java/com/emc/pravega/perf/PravegaWriterWorker.java (hunk 5 above)
Comment (kmgowda) [reply src/main/java/com/emc/pravega/perf/PravegaWriterWorker.java]: ok, i have removed them now

File: src/main/java/com/emc/pravega/perf/WriterWorker.java (hunk 7 above)
Comment (kmgowda) [reply src/main/java/com/emc/pravega/perf/WriterWorker.java]: do want me to remove ExecutionException?

File: src/main/java/com/emc/pravega/perf/WriterWorker.java (hunk 8 above)
Comment (kmgowda) [reply src/main/java/com/emc/pravega/perf/WriterWorker.java]: thanks.. yes , its done.

File: src/main/java/com/emc/pravega/perf/PravegaReaderWorker.java (hunk 11 above)
Comment (kmgowda) [reply src/main/java/com/emc/pravega/perf/PravegaReaderWorker.java]: removed these methods

File: src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java (hunk 14 above)
Comment (kmgowda) [reply src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java]: yes, but i will do as part of  : https://github.com/pravega/pravega-benchmark/issues/16

File: src/main/java/com/emc/pravega/perf/PravegaReaderWorker.java (hunk 15 above)
Comment (kmgowda) [reply src/main/java/com/emc/pravega/perf/PravegaReaderWorker.java]: yes, its removed now

File: src/main/java/com/emc/pravega/perf/WriterWorker.java (hunk 17 above)
Comment (kmgowda) [reply src/main/java/com/emc/pravega/perf/WriterWorker.java]: removed

File: src/main/java/com/emc/pravega/perf/ReaderWorker.java (hunk 16 above)
Comment (kmgowda) [reply src/main/java/com/emc/pravega/perf/ReaderWorker.java]: removed

File: src/main/java/com/emc/pravega/perf/WriterWorker.java (hunk 9 above)
Comment (kmgowda) [reply src/main/java/com/emc/pravega/perf/WriterWorker.java]: yes.. this was bothering me since quite some time.. I followed the kafka benchmarking tool logic, so , i have changed the eventspecsec to events per producer /consumer. this new logic eliminates the total count and threadsleeps.

File: src/main/java/com/emc/pravega/perf/WriterWorker.java (hunk 12 above)
Comment (kmgowda) [reply src/main/java/com/emc/pravega/perf/WriterWorker.java]: i have removed all except the writedata and event.. these abstract methods need to extended by derived classes; so i have to keep them public.

File: src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java (hunk 2 above)
Comment (tkaitchuck) [reply src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java]: The method can throw IllegalStateException without it being declared in the function signature.
https://www.geeksforgeeks.org/checked-vs-unchecked-exceptions-in-java/

File: src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java (hunk 4 above)
Comment (tkaitchuck) [reply src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java]: RuntimeException is fine, just be sure to pass `e` as the second arg to the constructor so that the trace and cause is not lost.

File: src/main/java/com/emc/pravega/perf/WriterWorker.java (hunk 7 above)
Comment (tkaitchuck) [reply src/main/java/com/emc/pravega/perf/WriterWorker.java]: All three of these do not need to be enumerated here.

File: src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java (hunk 3 above)
Comment (tkaitchuck) [reply src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java]: This doesn't fix the problem if transactionsPerCommit is 1.

File: src/main/java/com/emc/pravega/perf/PerfStats.java [hunk 18]
```diff
@@ -6,9 +6,9 @@
  * to you under the Apache License, Version 2.0 (the
  * "License"); you may not use this file except in compliance
  * with the License. You may obtain a copy of the License at
- *
+ * <p>
```
[the remark this answers is not in the document]
Comment (kmgowda) [reply src/main/java/com/emc/pravega/perf/PerfStats.java]: done

File: src/main/java/com/emc/pravega/perf/PravegaPerfTest.java [hunk 19]
```diff
@@ -6,9 +6,7 @@
  * to you under the Apache License, Version 2.0 (the
  * "License"); you may not use this file except in compliance
  * with the License. You may obtain a copy of the License at
- *
```
[the remark this answers is not in the document]
Comment (kmgowda) [reply src/main/java/com/emc/pravega/perf/PravegaPerfTest.java]: done

File: src/main/java/com/emc/pravega/perf/PravegaPerfTest.java [hunk 20]
```diff
@@ -70,17 +68,16 @@
     private static int producerCount = 0;
     private static int consumerCount = 0;
     private static int segmentCount = 0;
-    private static int eventsPerSec = 40;
-    private static int runtimeSec = 10;
+    private static int eventsPerWorker = 40;
```
[the remark this answers is not in the document]
Comment (kmgowda) [reply src/main/java/com/emc/pravega/perf/PravegaPerfTest.java]: i have changed this to 3000 by default;
idea here is , if the user has not supplied for how long he wants to execute the benchmarking then it will run for 24hrs or  writes/reads 3000 events by one producer/reader which ever is earlier/less.

File: src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java (hunk 2 above)
Comment (kmgowda) [reply src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java]: done

File: src/main/java/com/emc/pravega/perf/WriterWorker.java (hunk 7 above)
Comment (kmgowda) [reply src/main/java/com/emc/pravega/perf/WriterWorker.java]: since the call() invokes future.get , it requires  InterruptedException, ExecutionException to checked , so i have removed only IllegalStateException

File: src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java (hunk 4 above)
Comment (kmgowda) [reply src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java]: yes , its done

File: src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java (hunk 3 above)
Comment (kmgowda) [reply src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java]: ok, then for time being for this PR, I have done some this:
        synchronized (this) {
                if (transEventCount.incrementAndGet() >= transactionsPerCommit) {
                    transEventCount.set(0);
                    curTrans.commit();
                    if (!transaction.compareAndSet(curTrans, producer.beginTxn())) {
                        throw new IllegalStateException("WriteData called on the same PravegaTransactionWriterWorker from two threads in parallel.");
                    }
                }

synchronized count and commit.

while addressing https://github.com/pravega/pravega-benchmark/issues/16, I will revisit this logic to minimize the sync block

File: src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java (hunk 4 above)
Comment (kmgowda) [reply src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java]: done

File: src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java [hunk 21]
```diff
@@ -29,35 +29,39 @@
...
 
-        curTrans.writeEvent(key, data);
-        if (eventCount.incrementAndGet() >= transactionsPerCommit) {
-            eventCount.set(0);
-            curTrans.commit();
-            if (!transaction.compareAndSet(curTrans, producer.beginTxn())) {
-                throw new IllegalStateException("WriteData called on the same PravegaTransactionWriterWorker from two threads in parallel.");
+        try {
+            curTrans.writeEvent(key, data);
+            synchronized (this) {
```
Comment (tkaitchuck) [inline src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java]: This solves the race condition. Now you can simplify transEventCount can be a regular integer guardedBy("this"). And the if block that throws the illegal state can be removed.

Additionally you can also do the same with transaction. It can be made into a normal member guarded by 'this', and you don't need the atomic reference any more.

File: src/main/java/com/emc/pravega/perf/ReaderWorker.java [hunk 22]
```diff
@@ -0,0 +1,83 @@
...
+     * close the consumer/reader.
+     */
+    public abstract void close();
+
+    /**
+     * read all the data from reader/consumer.
+     *
+     * @param drainStats object which collects the details about the data read till there is no data to read from stream
+     */
+    public void cleanupEvents(PerfStats drainStats) {
```
Comment (tkaitchuck) [inline src/main/java/com/emc/pravega/perf/ReaderWorker.java]: It's weird that this is called 'cleanup' it doesn't actually result in anything being deleted. Maybe it should be called "readRemainingEvents" or something.

File: src/main/java/com/emc/pravega/perf/ReaderWorker.java [hunk 23]
```diff
@@ -0,0 +1,83 @@
...
+        } while (ret != null);
+        drainStats.printTotal(Instant.now());
+    }
+
+    @Override
+    public Void call() {
+        String ret = null;
+        try {
+
+            for (int i = 0; (i < eventsPerWorker) &&
```
Comment (tkaitchuck) [inline src/main/java/com/emc/pravega/perf/ReaderWorker.java]: Should readers be capped based on a count per worker? The events won't be uniformly distributed, so if some exit early because they hit the count it's going to mean some events go unread. Perhaps readers should just be a time thing, or perhaps they could have some sort of 'stop()' method. (If you want to do that in a different PR that's fine)

File: src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java (hunk 21 above)
Comment (kmgowda) [reply src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java]: yes , its done

File: src/main/java/com/emc/pravega/perf/ReaderWorker.java (hunk 22 above)
Comment (kmgowda) [reply src/main/java/com/emc/pravega/perf/ReaderWorker.java]: this cleanup gets invoked if both producer and consumers are specified with bench marking.. before the producer starts writing , this methods will clean the existing events. since its invoked before reading its kept as cleanupEvents and i am just maintain this method as it is.  i will revisit this soon.

File: src/main/java/com/emc/pravega/perf/ReaderWorker.java (hunk 23 above)
Comment (kmgowda) [reply src/main/java/com/emc/pravega/perf/ReaderWorker.java]: yes, currently I am just making sure that reader and transaction works fine but i am not publishing the performance numbers of reader and transactions.  once we close the pravega writer performance v/s kafka writer , I will revisit this code again.
